### ISSUE TITLE
Issue in building project for python 3.6

### ISSUE THREAD

**Issue by Dorak5** [Azure/azure-iot-sdk-python#120]
- **OS and version used:** <VERSION>macOS High Sierra 

- **Python runtime used:** <VERSION>3.6 

- **SDK version used:** <VERSION>release_2018_01_31 

# Description of the issue:

Every time I run the "./Build.sh --python-version 3.6" I get the following error message:

[ 51%] Built target iothub_client_amqp_ws_transport
make[1]: *** [python_service_client/tests/CMakeFiles/iothub_service_client_mock.dir/all] Error 2
make: *** [all] Error 2
Test project /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac
No tests were found!!!
~/projects/azure-iot-sdk-python
copy iothub_client library to samples folder
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/python/src/iothub_client.dylib: No such file or directory
copy iothub_client_mock library to tests folder
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/python/test/iothub_client_mock.dylib: No such file or directory
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/python/src/iothub_client.dylib: No such file or directory
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/python_service_client/src/iothub_service_client.dylib: No such file or directory
copy iothub_service_client library to samples folder
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/python_service_client/src/iothub_service_client.dylib: No such file or directory
copy iothub_service_client_mock library to tests folder
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/python_service_client/tests/iothub_service_client_mock.dylib: No such file or directory
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/python_service_client/src/iothub_service_client.dylib: No such file or directory
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/python/src/iothub_client.dylib: No such file or directory
copy provisioning_device_client library to samples folder
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/provisioning_device_client_python/src/provisioning_device_client.dylib: No such file or directory
copy provisioning_device_client_mock library to tests folder
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/provisioning_device_client_python/tests/provisioning_device_client_mock.dylib: No such file or directory
cp: /Users/<username>/projects/azure-iot-sdk-python/c/cmake/iotsdk_mac/provisioning_device_client_python/src/provisioning_device_client.dylib: No such file or directory
python3.6 iothub_client_ut.py
Traceback (most recent call last):
  File "iothub_client_ut.py", line 9, in <module>
    import iothub_client_mock
ModuleNotFoundError: No module named 'iothub_client_mock'

I've seen similar issues where a user didn't clone using --recursive or didn't use git submodule update --init --recursive.  I've done both and neither work.

The sdk does build for python 2.7.

# Code sample exhibiting the issue:

# Console log of the issue:

**Comment by zolvarga** [comment]
@Dorak5 

Please make sure after you cloned the azure-iot-sdk-python repo you run the following command:
git submodule update --init --recursive

Try to build again and let me know.

Best Regards,
Zoltan

**Comment by Dorak5** [comment]
Hi @zolvarga 

I cloned the azure-iot-sdk-python repo again using the command:
git clone --recursive https://github.com/Azure/azure-iot-sdk-python.git

While this was running I can see where its retrieving the submodules for c.  Once the clone was complete, I ran the command:
git submodule update --init --recursive

However, I didn't see this command load any additional submodules.  I then ran the following commands:
sh ./setup.sh --python-version 3.6
sh ./build.sh --build-python 3.6

Unfortunately, I still get the same error.

Hope this is helpful,
Mike D.

**Comment by zolvarga** [comment]
@Dorak5 

Thanks Mike, we will investigate it.

Best Regards,
Zoltan

**Comment by Dorak5** [comment]
Hi @zolvarga 

I haven't heard anything in a while.  Whats the good word?  :)

**Comment by tomafischer** [comment]
Same issues here. Any good news for us?

**Comment by alanjcastonguay** [comment]
I observe similar behaviour. Python 3.6.5 on Mac 10.13.4. `setup.sh` says boost 1.67.0, openssl 1.0.2o_1, cmake 3.11.0 installed. The `c` submodule is at commit 248ef1489b15532d2617c95e1feefecba85d88d2

The first erroneous part of `build.sh` output appears to be "-- Could NOT find Boost", originating from the call to `find_package(Boost COMPONENTS python)` on line 41. Adding `REQUIRED` yields

```
CMake Error at /usr/local/Cellar/cmake/3.11.0/share/cmake/Modules/FindBoost.cmake:2044 (message):
  Unable to find the requested Boost libraries.

  Boost version: 1.67.0

  Boost include path: /usr/local/include

  Could not find the following Boost libraries:

          boost_python

  No Boost libraries were found.  You may need to set BOOST_LIBRARYDIR to the
  directory containing Boost libraries or BOOST_ROOT to the location of
  Boost.
Call Stack (most recent call first):
  .../azure-iot-sdk-python/provisioning_device_client/CMakeLists.txt:41 (find_package)
```

So `boost` is being found, but `boost_python` is not. 

These exist
* `/usr/local/include/boost -> ../Cellar/boost/1.67.0/include/boost`
* `/usr/local/include/boost/python` (containing source)
* `/usr/local/Cellar/boost-python/1.67.0/` (containing dynlib with python27??)

This looks similar to https://github.com/openai/roboschool/issues/118

I note that `build_all/Mac/setup.sh`  runs `brew install boost-python --with-python3`, which emits the warning `this formula has no --with-python3 option so it will be ignored` and proceeds to install a boost-python for python27

Editing `build_all/Mac/setup.sh` to 
```diff
-       deps="boost-python --with-python3"
+       deps="boost-python3"
```

... and rerunning `./setup.sh --python-version 3.6` produces a `/usr/local/Cellar/boost-python3/1.67.0/lib/` containing `libboost_python36.dylib`.

But CMake refuses to find it. CMake is looking in `/usr/local/include/../lib`, which does contain a symlink `libboost_python36.dylib -> ../Cellar/boost-python3/1.67.0/lib/libboost_python36.dylib`.

Editing `device/iothub_client_python/CMakeLists.txt` and `provisioning_device_client/CMakeLists.txt` and `service/CMakeLists.txt` with

```
-        find_package(Boost COMPONENTS python)
+        string(REPLACE "." "" boost_python ${build_python})
+        find_package(Boost COMPONENTS python${boost_python})
```

... and rerunning `./build.sh --build-python 3.6` yields output

```
-- Boost version: 1.67.0
-- Found the following Boost libraries:
--   python36
```

... and the build completes and tests pass.

I captured these changes in https://github.com/alanjcastonguay/azure-iot-sdk-python/commit/9268ec5901b36d35a33b874380c39d7840a93a0a - @zolvarga what do you think?

**Comment by Xirui** [comment]
Facing the exactly same issue, applied the changes by @alanjcastonguay . It solved the problem.
Thanks @alanjcastonguay

**Comment by alanjcastonguay** [comment]
@Xirui thanks for the confirmation; glad it's working!

**Comment by CarlosLabrado** [comment]
This partially worked for me, I had to do a brew install boost-python3 for it to work.

**Comment by arthuston** [comment]
There looks to be a small typo in @alanjcastonguay's changes for setup.sh:
/usr/local/Cellar/boost-python/1.67.0/ (containing dynlib with python27??)

Which for me was 
/usr/local/Cellar/boost-python3/1.67.0

$ ls /usr/local/Cellar/boost-python3/1.67.0/lib
libboost_numpy36-mt.a		libboost_numpy36.a		libboost_python36-mt.a		libboost_python36.a
libboost_numpy36-mt.dylib	libboost_numpy36.dylib		libboost_python36-mt.dylib	libboost_python36.dylib

**Comment by alanjcastonguay** [comment]
@arthuston Thanks for the feedback. Where are you seeing the typo?

**Comment by Dorak5** [comment]
Hi @alanjcastonguay,

I applied your changes and ran into no issues building.  Thank you very much for the help.

**Comment by arthuston** [comment]
@alanjcastonguay Maybe not a typo, but perhaps just a difference in my results. Your post said /usr/local/Cellar/boost-python/1.67.0/ (containing dynlib with python27??) (e.g. "boost-python").

For me the location was /usr/local/Cellar/boost-python3/1.67.0 (eg. boost-python3)

**Comment by arthuston** [comment]
The typo I'm referring to is (e.g it should be boost-python3) was fixed in Merge pull request #3 from alanjcastonguay/fix-issue-120-python36-bui…  …

**Comment by arthuston** [comment]
I'm getting a second error. Here are my steps:
  548  git clone https://github.com/alanjcastonguay/azure-iot-sdk-python.git
  549  cd azure-iot-sdk-python/
  550  git submodule update --init --recursive
  551  cd build_all/mac
  552  ./build.sh --build-python 3.6
  553  ./setup.sh --python-version 3.6
  554  ./build.sh --build-python 3.6

And the build error
<snip>
[ 99%] Linking CXX shared library provisioning_device_client.dylib
ld: warning: -L path '/anaconda3/lib/libpython3.6m.dylib' is not a directory
[ 99%] Built target provisioning_device_client_python
[ 99%] Linking CXX shared library iothub_service_client.dylib
ld: warning: -L path '/anaconda3/lib/libpython3.6m.dylib' is not a directory
[ 99%] Built target iothub_service_client_python
[100%] Linking CXX shared library iothub_client.dylib
ld: warning: -L path '/anaconda3/lib/libpython3.6m.dylib' is not a directory
[100%] Built target iothub_client_python
Test project /Users/art/git/azure/azure-iot-sdk-python/c/cmake/iotsdk_mac
No tests were found!!!
~/git/azure/azure-iot-sdk-python
copy iothub_client library to samples folder
copy iothub_client_mock library to tests folder
copy iothub_service_client library to samples folder
copy iothub_service_client_mock library to tests folder
copy provisioning_device_client library to samples folder
copy provisioning_device_client_mock library to tests folder
python3.6 iothub_client_ut.py
Fatal Python error: take_gil: NULL tstate

./build.sh: line 96: 43612 Abort trap: 6           "python${PYTHON_VERSION}" iothub_client_ut.py

**Comment by arthuston** [comment]
build.sh fails when using anaconda python 3.6 to build the SDK. Using python 3.5 works.

$ conda install python=3.6
$ python3 --version
Python 3.6.5 :: Anaconda custom (64-bit)
$ ./setup.sh --python-version 3.5
$ ./build.sh --build-python 3.5
<snip>
python3.6 iothub_service_client_ut.py
Fatal Python error: take_gil: NULL tstate

./build.sh: line 105: 45028 Abort trap: 6           "python${PYTHON_VERSION}" iothub_service_client_ut.py

# this works
$ conda install python=3.5
$ python3 --version
Python 3.5.5 :: Anaconda custom (64-bit)
$ ./setup.sh --python-version 3.5
$ ./build.sh --build-python 3.5

**Comment by pierreca** [comment]
@Xirui @alanjcastonguay @Dorak5 @arthuston I tried to fix the mac build with #155, made it work with python 2.7 as well as 3.x up to 3.7. I think it's resolved so I'm closing this issue but feel free to reopen (or open a new one) if you run into other issues. Thanks!


*Thanks again for filing this issue! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://www.surveymonkey.com/r/Q5PLZ3V)*

**Comment by axper** [comment]
Install Mac OS X Command Line Tools first: http://osxdaily.com/2014/02/12/install-command-line-tools-mac-os-x/ . Spent 5 hours trying to figure this out, hope this helps someone.
